### DISCUSSION TITLE
How can I use my HTTPS domain point to my Supabase server?

### DISCUSSION THREAD

**Discussion by globemediaofficial** [supabase/supabase#10799] [Questions] [answered]
My goal is to be able to reach everything, including the studio, externally using my subdomain and `HTTPS`. There are a few tutorials for this using Nginx and Caddy but none have worked for me. I'll add what I have tried and the errors that occurred below. I'm open to ideas on how to fix it or ideas of how to do what I need differently.

### Basic subdomain from Ionos with SSL enabled
- Points to server IP address

### Caddy Configuration (generic version)
- Came from [this post](https://github.com/supabase/supabase/issues/10074#issuecomment-1340194036) and the linked guide
```
https://yourdomain.com {
        reverse_proxy /rest/v1/* localhost:8000
        reverse_proxy /auth/v1/* localhost:8000
        reverse_proxy /realtime/v1/* localhost:8000
        reverse_proxy /storage/v1/* localhost:8000
        reverse_proxy * localhost:3000
        basicauth / {
                yourusername $yourhashedpassword
        }
}
````
### If I have port forwarding enabled for `3000` and `8000` (want to avoid this)
- I can visit those ports directly from `http://yourdomain.com:3000` or `http://yourdomain.com:8000`
  - NOTE: This is only `HTTP` and does not work with `HTTPS` which I need for OAuth (see bottom of question)
### If there is not port forwarding (besides `80` and `443`)
- The reverse proxy does not work
  - I get `ERR_CONNECTION_TIMED_OUT`
### If I visit the IP directly
- Same as above with subdomain

## What I tried
- I tried with Nginx and ran into the same error
  - I thought it was a firewall issue but it persists, even when the firewall is disabled
  - It's possible that port `80` is not reachable but none of the guides showed steps for doing that. I enabled port forwarding for it and allow it in the firewall so I don't know what I am missing.

## Why I need `HTTPS`
- To setup Apple OAuth, and other providers, the redirect URL needs to be HTTPS

**Comment by humanshield89** [comment]
Does it not work like any other backend you are serving? 

In `Nginx`, to reverse proxy, a local port to `HTTPS` do something like this. 

## Assumptions
your domain: `domain.com`
server local services: `localhost:3000` and `8000`

Let's create a reverse proxy with HTTPS for the service at port 3000

1. in your `DNS`, create a subdomain like `api.domain.com` and point it to your server IP (only if you do not want to use the main domain for this service; otherwise, point domain.com to your IP)
2. now, on your server, add a new NJINX config like this
```nginx
server {
       server_name api.domain.com;

       location / {
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-NginX-Proxy true;
            # make sure this is the same as your port
            proxy_pass http://127.0.0.1:3000; # if your service is on a different port, change this 
            proxy_ssl_session_reuse off;
            proxy_set_header Host $http_host;
            proxy_cache_bypass $http_upgrade;
            proxy_redirect off;
       }

    listen [::]:443 ssl; 
    listen 443 ssl; 
    ssl_certificate /etc/letsencrypt/live/api.domain.com/fullchain.pem; # change to the path of your fullchain.pem
    ssl_certificate_key /etc/letsencrypt/live/api.domain.com/privkey.pem; # change to the path of your privkey.pem
    include /etc/letsencrypt/options-ssl-nginx.conf; 
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; 

}
server {
    if ($host = api.domain.com) {
        return 301 https://$host$request_uri;
    } 

       listen 80;
       listen [::]:80;
       server_name api.domain.com;
    return 404;
}
```
3. make sure to reload your Nginx configs
```bash
sudo systemctl reload nginx
```
4. if you visit api.domain.com you will find the service served on port 3000 on your server 
5. Do the same with other services, you have to map each port to a new subdomain to create reverse proxy `localPort` => `443`

**Comment by globemediaofficial** [comment, marked as answer]
I found something that somewhat worked, using the Nginx Docker image and Cloudflare Tunnels. With [this guide](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/Docker-Nginx-reverse-proxy-setup-example?amp=1), I am now able to see the Nginx welcome screen when I go to my subdomain, which is farther than I ever got before. Thanks to the Cloudflare Tunnel, it is automatically HTTPS. However, since Nginx is run through Docker in this case, I can’t figure out how to apple a reverse proxy. I still need help configuring the Nginx Docker image, because it doesn’t automatically work with the same configuration file that the regular Nginx did.

**Reply by globemediaofficial** [reply]
I figured out how to edit the configuration but now I get `502 Bad gateway`. This is the Nginx error:
```
2022/12/10 07:28:38 [error] 193#193: *27 connect() failed (111: Connection refused) while connecting to upstream, client: 172.19.0.2, server: localhost,, request: "GET / HTTP/1.1", upstream: "http://0.0.0.0:3000/", host: "api.example.com"
172.19.0.2 
```

**Comment by humanshield89** [comment]
can you share your Nginx config file? And are you sure a service is running at port 3000 and Nginx can access it? You have your certificates in place, if you have cloud flare, you can use a self-signed certificate with a cloud flare SSL

**Reply by globemediaofficial** [reply]
@humanshield89 @kiwicopple 
> are you sure a service is running at port 3000 and Nginx can access it

I'm sure it is running on port 3000, because if I directly visit `0.0.0.0:3000` it works. I'm not sure if Nginx can access it, but I don't see why it wouldn't be able to. The error says connection refused on `0.0.0.0:3000` but I don't know why.

>  You have your certificates in place, if you have cloud flare, you can use a self-signed certificate with a cloud flare SSL

I'm not even trying to use `SSL` in Nginx. My Tunnel is set up to go from `https://api.example.com` to `http://nginx:80` which is all I need. My current config is from a guide so I don't know which lines I can remove to get rid of the `SSL` and `connection_upgrade`.

## My Nginx Config
```
map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
}

upstream supabase {
    server 0.0.0.0:3000;
}

upstream kong {
    server 0.0.0.0:8000;
}

server {
    listen       80;
    server_name  localhost, 0.0.0.0, api.example.com;

    #access_log  /var/log/nginx/host.access.log  main;

    
    # REST API
    location ~ ^/rest/v1/(.*)$ {
        proxy_set_header Host $host;
        proxy_pass http://kong;
        proxy_redirect off;
    }

    # Authentication
    location ~ ^/auth/v1/(.*)$ {
        proxy_set_header Host $host;
        proxy_pass http://kong;
        proxy_redirect off;
    }

    # Realtime
    location ~ ^/realtime/v1/(.*)$ {
        proxy_redirect off;
        proxy_pass http://kong;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
        proxy_set_header Host $host;
    }
    
    # Studio
    location / {
        proxy_set_header Host $host;
        proxy_pass http://supabase;
        proxy_redirect off;
        proxy_set_header Upgrade $http_upgrade;
    }
}
```

**Comment by ymerdrengene** [comment]
@globemediaofficial did u manage to work=
